Transport/Measure-EmailDelayInMTL.ps1

```diff
@@ -0,0 +1,172 @@
+    Return $true
+}
+
+if (Test-ScriptVersion -AutoUpdate) {
+    # Update was downloaded, so stop here.
+    Write-Host "Script was updated. Please rerun the command."
+    return
+}
+
+# make sure out output variable is null
+$output = $Null
+
+# Test for the provided file and load it.
+if (Test-Path $MTLFile) {
+    Write-Output "Loading MTL file."
+    $mtl = Import-Csv $MTLFile -Encoding Unicode
+} else {
+    Write-Error "Unable to find specified file"
+}
+
+# Validate the MTL
+if (Test-CSVData -CSV $mtl -ColumnsToCheck "event_id", "source", "message_id", "date_time_utc") {}
+else {
+    Write-Error "MTL is missing one or more required fields: `"event_id`",`"source`",`"message_id`",`"date_time_utc`"" -ErrorAction Stop
+}
+
+# get all of the unique message IDs in the file.
+[array]$uniqueMessageIDs = $mtl | Select-Object -ExpandProperty message_id | Sort-Object | Get-Unique
+
+if ($uniqueMessageIDs.count -eq 0) {
+    Write-Error "No Unique MessageIDs found in data."
+}
+
+# Carve the data up into smaller collections to make searching faster.
+# Most of what is in the MTL we don't need for this.
+$SMTPReceive = $mtl | Where-Object { ($_.event_id -eq 'RECEIVE') -and ($_.source -eq 'SMTP') }
+$StoreDeliver = $mtl | Where-Object { ($_.event_id -eq 'DELIVER') -and ($_.source -eq 'STOREDRIVER') }
+$SMTPDeliver = $mtl | Where-Object { ($_.event_id -eq 'SENDEXTERNAL') -and ($_.source -eq 'SMTP') }
+
+# Loop thru each unique messageID
+foreach ($id in $uniqueMessageIDs) {
+
+    # make sure we aren't carrying anything over from the previous foreach.
+    $AllSentTimes = $Null
+    $AllStoreDeliverTimes = $Null
+    $AllRemoteDeliverTimes = $Null
+
+    # extract the times for a message ID ... there can be more than one of each of these.
+    [array]$AllSentTimes = ($SMTPReceive | Where-Object { ($_.message_id -eq $id) }).date_time_utc
+    [array]$AllStoreDeliverTimes = ($StoreDeliver | Where-Object { ($_.message_id -eq $id) }).date_time_utc
+    [array]$AllRemoteDeliverTimes = ($SMTPDeliver | Where-Object { ($_.message_id -eq $id) }).date_time_utc
+
+    # Process the time sent
+    if ($AllSentTimes.count -eq 0) {
+        Write-Warning ($id.message_id.ToString() + "unable to find sent time. Discarding messageID")
+        quit
+    } else {
+        $TimeSent = Get-Date ($AllSentTimes | Sort-Object | Select-Object -First 1)
+    }
+
+    # If we didn't find any delivery information then drop the message ID
+    if ($AllStoreDeliverTimes.count -eq 0 -and $AllRemoteDeliverTimes.count -eq 0) {
+        Write-Warning ($id + "not able to find delivery time in MTL. Discarding messageID")
+    }
+    # Process the message information
+    else {
+
+        # Combine all of the delivery times.
+        [array]$AllDeliveries = ($AllStoreDeliverTimes + $AllRemoteDeliverTimes) | Sort-Object
+
+        $report = [PSCustomObject]@{
+            MessageID    = $id
+            TimeSent     = $TimeSent
+            TimeReceived = (Get-Date $AllDeliveries[-1])
```

Comment:
You are assuming that the time is in order. How do we know this is ALWAYS true?

Reply:
not assuming that ... the array is sorted before we pull from it.
But I moved the logic around a bit and changed the variable names to call out that they are sorted.